lib/urlapi.c

```diff
@@ -1102,10 +1102,10 @@ static CURLUcode parseurl(const char *url, CURLU *u, unsigned int flags)
     hostp = p; /* host name starts here */
 
     /* find the end of the host name + port number */
-    while(*p && !HOSTNAME_END(*p))
+    while(p && *p && !HOSTNAME_END(*p))
```

Comment:
How can `p` be NULL here?

Reply:
If schemelen == false, p can be not initialized...

Reply:
Maybe need also 'const char *p = NULL;'?

Reply:
Isn't `p` then assigned on line 1100?

Reply:
Oh, see...
But url can be NULL? On release build checking `DEBUGASSERT(url)` is dummy and potential p can be NULL?

Reply:
`url` should not be possible to be NULL as input, the assert is there to make sure it dies if that would ever happen anyway.

Reply:
```c
#if defined(DEBUGBUILD)
#define DEBUGASSERT(x) assert(x)
#else
#define DEBUGASSERT(x) do { } while(0)
#endif
```
assert enabled only for DEBUGBUILD, and on release build if potentially `url == NULL` the application dies, or not?

Reply:
Then how does `url` end up NULL  there?

Reply:
Someone will write a code with an bug that will result in `parseurl` call with `url == NULL`.

Reply:
The ASSERT is there to catch such possible future mistake (in tests). `parseurl()` is only called from within a single function: `parseurl_and_replace()`.

`parseurl_and_replace()` is called on two places and the URL argument cannot be NULL in either invocation.

If `parseurl()` needs to handle a NULL url then that check should be added explicitly in the beginning of the function. But I don't think it needs to.

Reply:
Ok, i get it.